### ISSUE TITLE
interact_by_users function not used in follow_user_followers?

### ISSUE THREAD

**Issue by xvx29a** [InstaPy/InstaPy#1983]
## Current Behavior
During follow_user_followers shouldn't interact_by_users be called instead of like_by_users in the interact: section? The like_by_users is trying to follow users that have already been followed from the follow_user_followers function. This causes an issue if you are trying to avoid private users, like from #1466 

 ```

        if interact:
            self.logger.info('--> User followed: {}'.format(userFollowed))
            userFollowed = random.sample(userFollowed, int(ceil(
                self.user_interact_percentage * len(userFollowed) / 100)))
            self.like_by_users(userFollowed,
                               self.user_interact_amount,
                               self.user_interact_random,
                               self.user_interact_media)

        return self
```
to
```

        if interact:
            self.logger.info('--> User followed: {}'.format(userFollowed))
            userFollowed = random.sample(userFollowed, int(ceil(
                self.user_interact_percentage * len(userFollowed) / 100)))
            self.interact_by_users(userFollowed,
                               self.user_interact_amount,
                               self.user_interact_random,
                               self.user_interact_media)

        return self
```

interact_by_users actually doesn't look like it's used anywhere.

**Comment by xvx29a** [comment]
Just saw that this was a new function added by @uluQulu. follow_user_followers looks like the perfect spot for that function.

**Comment by uluQulu** [comment]
Hey @xvx29a,
That great function, `interact_by_users` is written by somebody else but not me :)
Also, `interact_by_users` function also _tries to_ follow a user, if `self.do_follow` is _enabled_...
But I like your idea, it can be improved, by adding `self.do_follow=False` before it and `self.do_follow=True` after it
```
original_do_follow = self.do_follow
self_do_follow = False
self.like_by_users(userFollowed,
                   self.user_interact_amount,
                   self.user_interact_random,
                   self.user_interact_media)
self.do_follow = original_do_follow
```
I will use this idea in my current PR, too, where it was trying to follow with `interact_by_users` even after it followed a user with `follow_by_list` feature...  
Thanks for reminding me this great idea!
Also, it would be great if you finalize your modifications a PR the changes.
Good luck!

**Comment by xvx29a** [comment]
Thanks! I'll have to add that in. 

There is an issue with the interact_by_users function with the new relationship ratio, not sure if it also happens with like_by_users.  This is happening with follow_user_followers with interact_by_users replaced and interact set to true.

```
INFO [2018-05-08 08:37:08] [xxxxxxxx]  Username [2/4]
INFO [2018-05-08 08:37:08] [xxxxxxxx]  --> b'ig_user'
INFO [2018-05-08 08:37:08] [xxxxxxxx]  Getting ig_user image list...
INFO [2018-05-08 08:37:34] [xxxxxxxx]  Interaction begin...
INFO [2018-05-08 08:38:06] [xxxxxxxx]  Post [1/2]
INFO [2018-05-08 08:38:06] [xxxxxxxx]     https://www.instagram.com/p/rand/?taken-by=ig_user
INFO [2018-05-08 08:38:09] [xxxxxxxx]  Image from: b'ig_user'
INFO [2018-05-08 08:38:12] [xxxxxxxx]  User: ig_user >> followers: 7568  |  following: 1208  |  relationship ratio: 6.26
INFO [2018-05-08 08:38:12] [xxxxxxxx]  --> Image not liked: b"User ig_user's followers count exceeds maximum limit"
INFO [2018-05-08 08:38:12] [xxxxxxxx]  Post [1/2]
INFO [2018-05-08 08:38:12] [xxxxxxxx]   https://www.instagram.com/p/rand/?taken-by=ig_user
INFO [2018-05-08 08:38:15] [xxxxxxxx]  Image from: b'ig_user'
INFO [2018-05-08 08:38:19] [xxxxxxxx]  User: ig_user >> followers: 7568  |  following: 1208  |  relationship ratio: 6.26
INFO [2018-05-08 08:38:19] [xxxxxxxx]  --> Image not liked: b"User ig_user's followers count exceeds maximum limit"
INFO [2018-05-08 08:38:19] [xxxxxxxx]  Post [1/2]
INFO [2018-05-08 08:38:19] [xxxxxxxx]   https://www.instagram.com/p/rand/?taken-by=ig_user
INFO [2018-05-08 08:38:21] [xxxxxxxx]  Image from: b'ig_user'
INFO [2018-05-08 08:38:25] [xxxxxxxx]  User: ig_user >> followers: 7568  |  following: 1208  |  relationship ratio: 6.26
INFO [2018-05-08 08:38:25] [xxxxxxxx]  --> Image not liked: b"User ig_user's followers count exceeds maximum limit"
INFO [2018-05-08 08:38:25] [xxxxxxxx]  Post [1/2]
INFO [2018-05-08 08:38:25] [xxxxxxxx]    https://www.instagram.com/p/rand/?taken-by=ig_user
INFO [2018-05-08 08:38:27] [xxxxxxxx]  Image from: b'ig_user'
INFO [2018-05-08 08:38:31] [xxxxxxxx]  User: ig_user >> followers: 7568  |  following: 1208  |  relationship ratio: 6.26
INFO [2018-05-08 08:38:31] [xxxxxxxx]  --> Image not liked: b"User ig_user's followers count exceeds maximum limit"
INFO [2018-05-08 08:38:31] [xxxxxxxx]  Post [1/2]
INFO [2018-05-08 08:38:31] [xxxxxxxx]    https://www.instagram.com/p/rand/?taken-by=ig_user
INFO [2018-05-08 08:38:33] [xxxxxxxx]  Image from: b'ig_user'
INFO [2018-05-08 08:38:36] [xxxxxxxx]  User: ig_user >> followers: 7568  |  following: 1208  |  relationship ratio: 6.26
INFO [2018-05-08 08:38:36] [xxxxxxxx]  --> Image not liked: b"User ig_user's followers count exceeds maximum limit"
INFO [2018-05-08 08:38:36] [xxxxxxxx]  Post [1/2]
INFO [2018-05-08 08:38:36] [xxxxxxxx]    https://www.instagram.com/p/rand/?taken-by=ig_user
INFO [2018-05-08 08:38:38] [xxxxxxxx]  Image from: b'ig_user'
INFO [2018-05-08 08:38:41] [xxxxxxxx]  User: ig_user >> followers: 7568  |  following: 1208  |  relationship ratio: 6.26
INFO [2018-05-08 08:38:41] [xxxxxxxx]  --> Image not liked: b"User ig_user's followers count exceeds maximum limit"
INFO [2018-05-08 08:38:41] [xxxxxxxx]  Post [1/2]
INFO [2018-05-08 08:38:41] [xxxxxxxx]    https://www.instagram.com/p/rand/?taken-by=ig_user
INFO [2018-05-08 08:38:44] [xxxxxxxx]  Image from: b'ig_user'
INFO [2018-05-08 08:38:48] [xxxxxxxx]  User: ig_user >> followers: 7568  |  following: 1208  |  relationship ratio: 6.26
INFO [2018-05-08 08:38:48] [xxxxxxxx]  --> Image not liked: b"User ig_user's followers count exceeds maximum limit"
INFO [2018-05-08 08:38:48] [xxxxxxxx]  Post [1/2]
INFO [2018-05-08 08:38:48] [xxxxxxxx]    https://www.instagram.com/p/rand/?taken-by=ig_user
INFO [2018-05-08 08:38:50] [xxxxxxxxx]  Image from: b'ig_user'
INFO [2018-05-08 08:38:54] [xxxxxxxxx]  User: ig_user >> followers: 7568  |  following: 1208  |  relationship ratio: 6.26
INFO [2018-05-08 08:38:54] [xxxxxxxxx]  --> Image not liked: b"User ig_user's followers count exceeds maximum limit"
INFO [2018-05-08 08:38:54] [xxxxxxxxx]  Post [1/2]
INFO [2018-05-08 08:38:54] [xxxxxxxxx]    https://www.instagram.com/p/rand/?taken-by=ig_user
INFO [2018-05-08 08:38:56] [xxxxxxxxx]  Image from: b'ig_user'
INFO [2018-05-08 08:38:59] [xxxxxxxxx]  User: ig_user >> followers: 7568  |  following: 1208  |  relationship ratio: 6.26
INFO [2018-05-08 08:38:59] [xxxxxxxxx]  --> Image not liked: b"User ig_user's followers count exceeds maximum limit"
INFO [2018-05-08 08:38:59] [xxxxxxxxx]  Post [1/2]
INFO [2018-05-08 08:38:59] [xxxxxxxxx]    https://www.instagram.com/p/rand/?taken-by=ig_user
INFO [2018-05-08 08:39:01] [xxxxxxxxx]  Image from: b'ig_user'
INFO [2018-05-08 08:39:04] [xxxxxxxxx]  User: ig_user >> followers: 7568  |  following: 1208  |  relationship ratio: 6.26
INFO [2018-05-08 08:39:04] [xxxxxxxxx]  --> Image not liked: b"User ig_user's followers count exceeds maximum limit"
INFO [2018-05-08 08:39:04] [xxxxxxxxx]  -------------
INFO [2018-05-08 08:39:04] [xxxxxxxxx]  --> Given amount not fullfilled, image pool reached its end
```

**Comment by uluQulu** [comment]
Yes I noticed that issue previously and included it's fix in my latest PR xD
The thing is your set interact amount to `2` and `randomize=True`
```python
session.set_user_interact(amount=2, randomize=True)
```
And while getting links for `interact_by_users` method in `get_links_for_username` function inside **like_util.py** file, it multiplies that amount (_`2`_) with `5` in case fo `randomize==True`
```powershell
if randomize:
    # Expanding the pooulation for better random distribution
    amount = amount * 5
```
_so your amount `2` became `10`  and then it will `return links[:amount]` so 10 links.._
And why it always shows `Post [1/2]` is it shows `[liked_img+1/amount]` and since no pic is liked due to **relationship bounds**, `liked_img` will be `0` and amount is your original amount here- `2`
....
**To solve** it, apply the change I have done in my latest PR, #1981:
in `interact_by_users` method of **instapy.py** file, find and **replace**:
```python
for i, link in enumerate(links):
```
**with**
```python
for i, link in enumerate(links[:amount]):
```
_now it should behave as expected..._

**Comment by xvx29a** [comment]
Another solution to this would be to build a user array in follow_user_followers from the dialog and then send it off to interact_by_users or follow_by_users, the same as your follow_user_likers.  Instead of using the follow_through_dialog(). Then the follow/interact can all be done via follow_by_users.

**Comment by uluQulu** [comment]
I think there is not need for it, but the idea you told is what I do think for the last couple of days, but for different reason raised by @Dimlsaev, https://github.com/timgrossmann/InstaPy/issues/1858#issuecomment-386646910
It will help us verify users through relationship bounds by following them one by one instead of inside dialog box.
> You seems to be good at reading and understanding source codes, just wonders, in which PL do you write the most? xD

**Comment by xvx29a** [comment]
I am really digging into this because I am trying to limit encounters with private users. So when I implemented your changes from #1466 it sent me down a rabbit hole chasing other bugs..

> I write mainly bash scripts, but have done php/perl/python
